lib/src/leak_tracking/retaining_path/_retaining_path.dart

```diff
@@ -0,0 +1,145 @@
+
+Future<void> _connect() async {
+  if (_connected) return;
+
+  final info = await Service.getInfo();
+  if (info.serverWebSocketUri == null) {
+    throw StateError(
+      'Leak troubleshooting is not available in release mode. Run your application or test with flag "--debug" '
+      '(Not supported for Flutter yet: https://github.com/flutter/flutter/issues/127331).',
+    );
+  }
+
+  _service = await connectWithWebSocket(info.serverWebSocketUri!, (error) {
+    throw error ?? Exception('Error connecting to service protocol');
+  });
+  await _service.getVersion();
+  await _getIdForTwoIsolates();
+
+  _connected = true;
+}
+
+/// Tries to wait for two isolates to be available.
+///
+/// Depending on environment (command line / IDE, Flutter / Dart), isolates may have different names,
+/// and there can be one or two. Sometimes the second one appears with latency.
+/// And sometimes there are two isolates with name 'main'.
+Future<void> _getIdForTwoIsolates() async {
+  _log.info('Waiting for two isolates to be available.');
+  const isolatesToGet = 2;
+  const watingTime = Duration(seconds: 2);
+  final stopwatch = Stopwatch()..start();
+  while (_isolateIds.length < isolatesToGet && stopwatch.elapsed < watingTime) {
+    _isolateIds.clear();
+    await forEachIsolate(
+      _service,
+      (IsolateRef r) async => _isolateIds.add(r.id!),
+    );
+    if (_isolateIds.length < isolatesToGet) {
+      await Future.delayed(const Duration(milliseconds: 100));
+    }
+  }
+  if (_isolateIds.isEmpty) {
+    throw StateError('Could not connect to isolates.');
+  }
+  _log.info('Number of isolates: ${_isolateIds.length}');
+}
+
+class _ObjectFingerprint {
+  _ObjectFingerprint(this.type, this.code);
+
+  final Type type;
+  final int code;
+}
+
+Future<_ItemInIsolate?> _objectInIsolate(_ObjectFingerprint object) async {
+  final classes = await _findClasses(object.type.toString());
+
+  for (final theClass in classes) {
+    final instances = (await _service.getInstances(
+          theClass.isolateId,
+          theClass.itemId,
+          10000000,
+        ))
+            .instances ??
+        <ObjRef>[];
+
+    final result = instances.firstWhereOrNull(
+      (objRef) =>
+          objRef is InstanceRef && objRef.identityHashCode == object.code,
+    );
+    if (result != null) {
+      return _ItemInIsolate(isolateId: theClass.isolateId, itemId: result.id!);
+    }
+  }
+
+  return null;
+}
+
+class _ItemInIsolate {
+  _ItemInIsolate({required this.isolateId, required this.itemId});
+
+  final String isolateId;
+  final String itemId;
+}
+
+Future<List<_ItemInIsolate>> _findClasses(String runtimeClassName) async {
+  final result = <_ItemInIsolate>[];
+
+  for (final isolateId in _isolateIds) {
+    var classes = await _service.getClassList(isolateId);
+
+    // In the beginning list of classes may be empty.
+    while (classes.classes?.isEmpty ?? true) {
+      await Future.delayed(const Duration(milliseconds: 100));
```

Comment:
we should have a timeout here so that this will not loop forever if for some reason no classes populate